Report unallocated memory only

kashizui opened the pull request:
First pass solution to #584 
Have workers report only allocated memory, while defining the potential memory usage of runs that don't have request_memory specified to be infinite.

Essentially implements:

jobs without request_memory:
* min: 0
* max: infty

jobs with request_memory:
* min: request_memory
* max: request_memory

```
for each job:
    for each worker:
        if worker.memory >= job.min_memory:
            worker.memory = max(0, worker.memory - job.max_memory)
            schedule(job, worker)
            break
``` 

One remaining problem here is that "free-for-all" jobs will still get scheduled on workers that already have "explicit" jobs (#584). To keep it simple, can fix this with appropriate if-statements in the scheduler code.

kashizui committed:
Subtract allocated memory

fabeschan commented:
Hmm lemme think if there is a better way to handle this "free-for-all" thing

fabeschan commented:
Hmm if I understand your pseudo code correctly, a single free-for-all job will reduce the worker's memory to 0?

percyliang commented:
But a free-for-all job can still run on that worker.

On Thu, Oct 19, 2017 at 1:10 AM, Naph <notifications@github.com> wrote:

> Hmm if I understand your pseudo code correctly, a single free-for-all job
> will reduce the worker's memory to 0?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/800#issuecomment-337832546>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAakuAMfHoZjtx-1wzO4mMHbAtjzam7Oks5stwP9gaJpZM4P5y_j>
> .
>

kashizui commented:
That’s right, once a free-for-all job gets assigned to a worker, its unallocated memory goes to 0, and then only free-for-all jobs can be further assigned to that worker.

fabeschan commented:
Ah, I see

percyliang commented:
What's the status of this?

kashizui commented:
This change is strictly better than what we had before (i.e. paying no attention to memory allocation), but still doesn't fully solve the problem of "guaranteeing X bytes of memory when a user requests X bytes of memory".

I just figured out a way for me to continue contributing to this patch through my employer's open-source policies. So I'm open to just merging this in or letting things simmer until I can come up with something better.

percyliang commented:
What about checking this in right now as a temporary solution and
submitting another PR if you happen to think of something?

On Mon, Oct 30, 2017 at 11:18 PM, Stephen Koo <notifications@github.com>
wrote:

> This change is strictly better than what we had before (i.e. paying no
> attention to memory allocation), but still doesn't fully solve the problem
> of "guaranteeing X bytes of memory when a user requests X bytes of memory".
>
> I just figured out a way for me to continue contributing to this patch
> through my employer's open-source policies. So I'm open to just merging
> this in or letting things simmer until I can come up with something better.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/800#issuecomment-340671154>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAakuNNAZWqmBTWy9rMwWOpIY3aAplLTks5sxru4gaJpZM4P5y_j>
> .
>

kashizui committed:
Don't fail jobs that request a lot of memory

This check no longer applies because worker memory now represents the
amount of *unallocated* memory, not the total memory available.
Instead, bundles will now just wait until a worker has enough
memory available.

kashizui commented:
I see that master's build is also failing.. is this still alright to merge?

percyliang commented:
Yeah...go ahead and merge...  we'll fix master.

On Tue, Nov 7, 2017 at 10:00 PM, Stephen Koo <notifications@github.com>
wrote:

> I see that master's build is also failing.. is this still alright to merge?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/800#issuecomment-342719160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAakuCoZ1ulaIziUWv5DQGTNhEEml7_hks5s0UNwgaJpZM4P5y_j>
> .
>
